File: 0015-3sum/0015-3sum.py
from collections import Counter
from itertools import combinations


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        triplets = []
        counter = Counter(nums)
        for n1, n2 in combinations(counter.keys(), 2):
            n3 = -(n1 + n2)
            if n1 < n3 and n2 < n3 and n3 in counter:
                triplets.append([n1, n2, n3])
        if counter.get(0, 0) >= 3:
            triplets.append([0, 0, 0])
        del counter[0]
        for n, count in counter.items():
            if count >= 2 and -2 * n in counter:
                triplets.append([n, n, -2 * n])
        return triplets

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        int n = nums.size();
        int left, right, sum;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < n; ++i) {
            if (nums[i] > 0)
                break;
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            left = i + 1;
            right = n - 1;
            while (left < right) {
                sum = nums[i] + nums[left] + nums[right];

                if (sum == 0) {
                    result.push_back({nums[i], nums[left], nums[right]});
                    left += 1;
                    right -= 1;
                    while (left < right && nums[left] == nums[left - 1]) {
                        left += 1;
                    }
                    while (left < right && nums[right] == nums[right + 1]) {
                        right -= 1;
                    }
                } else if (sum < 0) {
                    left += 1;
                } else {
                    right -= 1;
                }
            }
        }
        return result;
    }
};
